modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/GridSubqueryJoinOptimizer.java

```diff
@@ -0,0 +1,869 @@
+        if (targetEl != null)
+            targetEl.child(childInd, existsExpr);
+        else
+            parent.where(existsExpr);
+
+        boolean res = pullOutSubQryFromExistsClause(parent, targetEl, childInd);
+
+        if (!res) { // restore original query in case of failure
+            if (targetEl != null)
+                targetEl.child(childInd, oldInExpr);
+            else
+                parent.where(oldInExpr);
+
+            subS.where(oldCond);
+        }
+
+        return res;
+    }
+
+    /**
+     * Creates tree from provided elements which will be connected with an AND operator.
+     *
+     * @param ops Ops.
+     * @return Root of the resulting tree.
+     */
+    private static GridSqlElement buildConditionBush(List<GridSqlElement> ops) {
+        assert !F.isEmpty(ops);
+
+        if (ops.size() == 1)
+            return ops.get(0);
+
+        int m = (ops.size() + 1) / 2;
+
+        GridSqlElement left = buildConditionBush(ops.subList(0, m));
+        GridSqlElement right = buildConditionBush(ops.subList(m, ops.size()));
+
+        return new GridSqlOperation(AND, left, right);
+    }
+
+    /**
+     * Pull out sub-select from EXISTS clause to the parent select level.
+     * <p>
+     * Example:
+     * <pre>
+     *   Before:
+     *     SELECT name
+     *       FROM emp e
+     *      WHERE EXISTS (SELECT 1 FROM dep d WHERE d.id = e.dep_id and d.name = 'dep1')
+     *        AND sal > 2000
+     *
+     *   After:
+     *     SELECT name
+     *       FROM emp e
+     *       JOIN dep d
+     *      WHERE sal > 2000 AND d.id = e.dep_id and d.name = 'dep1
+     * </pre>
+     *
+     * @param parent Parent select.
+     * @param targetEl Target sql element. Can be null.
+     * @param childInd Column ind.
+     */
+    private static boolean pullOutSubQryFromExistsClause(
+        GridSqlSelect parent,
+        @Nullable GridSqlAst targetEl,
+        int childInd
+    ) {
+        // extract sub-query
+        GridSqlSubquery subQry = targetEl != null
+            ? targetEl.child(childInd).child()
+            : parent.where().child();
+
+        GridSqlSelect subS = checkSubQueryCanBePulledOut(subQry);
+        if (subS == null)
+            return false;
+
+        if (targetEl != null)
+            targetEl.child(childInd, subS.where());
+        else
+            parent.where(subS.where());
+
+        GridSqlElement parentFrom = parent.from() instanceof GridSqlElement
+            ? (GridSqlElement)parent.from()
+            : new GridSqlSubquery((GridSqlQuery)parent.from());
+
+        parent.from(new GridSqlJoin(parentFrom, GridSqlAlias.unwrap(subS.from()), false, null)).child();
+
+        return true;
+    }
+
+    /**
+     * Whether sub-query can be pulled out or not.
+     *
+     * @param subQry Sub-query.
+     * @return Sub-query if it can be pulled out or {@code null}.
+     */
+    private static GridSqlSelect checkSubQueryCanBePulledOut(GridSqlSubquery subQry) {
```

Comment:
Should be named trySubQueryPullOut or should return boolean/Boolean. Same with JavaDoc - its about possibility of doing something, but actually do it and return the result.